Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm

```diff
@@ -1354,19 +1368,72 @@ static bool isContextMenuEvent(const WebMouseEvent& event)
 void UnifiedPDFPlugin::scrollToPDFDestination(PDFDestination *destination)
 {
     auto unspecifiedValue = get_PDFKit_kPDFDestinationUnspecifiedValue();
+
+    auto pageIndex = [m_pdfDocument indexForPage:[destination page]];
     auto pointInPDFPageSpace = [destination point];
     if (pointInPDFPageSpace.x == unspecifiedValue)
         pointInPDFPageSpace.x = 0;
     if (pointInPDFPageSpace.y == unspecifiedValue)
-        pointInPDFPageSpace.y = 0;
+        pointInPDFPageSpace.y = pageHeight(pageIndex);
 
-    scrollToPointInPDF(roundedIntPoint(pointInPDFPageSpace), [m_pdfDocument indexForPage:[destination page]] - 1);
+    scrollToPointInPage(roundedIntPoint(pointInPDFPageSpace), pageIndex);
 }
 
-void UnifiedPDFPlugin::scrollToPointInPDF(IntPoint pointInPDFPageSpace, PDFDocumentLayout::PageIndex pageIndex)
+void UnifiedPDFPlugin::scrollToPointInPage(IntPoint pointInPDFPageSpace, PDFDocumentLayout::PageIndex pageIndex)
 {
-    auto pointInDocumentSpace = convertFromPageToContents(pointInPDFPageSpace, pageIndex);
-    scrollToPositionWithoutAnimation(pointInDocumentSpace);
+    auto pointInContentsSpace = convertFromPageToContents(pointInPDFPageSpace, pageIndex);
+    scrollToPositionWithoutAnimation(pointInContentsSpace);
+}
+
+void UnifiedPDFPlugin::scrollToPage(PDFDocumentLayout::PageIndex pageIndex)
+{
+    scrollToPointInPage({ 0, static_cast<int>(pageHeight(pageIndex)) }, pageIndex);
+}
+
+void UnifiedPDFPlugin::scrollToFragmentIfNeeded()
+{
+    if (!m_pdfDocument || !m_didAttachScrollingTreeNode)
+        return;
+
+    if (m_didScrollToFragment)
+        return;
+
+    m_didScrollToFragment = true;
+
+    if (!m_frame)
+        return;
+
+    auto fragment = m_frame->url().fragmentIdentifier();
+    if (!fragment)
+        return;
+
+    auto fragmentView = StringView(fragment);
+
+    // Only respect the first fragment component.
+    if (auto endOfFirstComponentLocation = fragmentView.find('&'); endOfFirstComponentLocation != notFound)
```

Comment:
Are there no helpers for these? Or are they PDF specific?

Reply:
We have query string parsers, this is like query format in the fragment. It's derived indirectly from the PDF spec which I guess is a bit more restrictive than a naive impl would be 🤷‍♂️